### PR TITLE
new API endpoint for propolis-server to replace a crucible downstairs

### PR DESCRIPTION
**Pull request by leftwo**
A new propolis-server API was added to support the replacement of a downstairs via a new VolumeConstructionRequest.  To do this replacement, we need the current VCR to go along with the new VCR, and we can get that from the InstanceSpec.

As part of this change we have moved the VmObjects InstanceSpec behind a Mutex, as a successful replacement of a crucible backends VCR should be followed by a replacement of the VCR in the instance spec.

Added a new vcr option to crucible-cli which can call this API

This also includes an updated Crucible, as we need the latest version that does not require a Logger with calls to target replace.

### CODE REVIEW

File: bin/propolis-server/src/lib/server.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/server.rs:978]: I think we need a better story for mutating a VM this way, but don't want to hold up progress on the replacement workflow while we work out exactly what that should look like. #205 is tracking the general "change a VM's configuration at runtime" story, so I suggest we refer to that here, and I'll bump its priority up accordingly.

```suggestion

    // TODO(#205): Mutating a VM's configuration should be a first-class
    // operation in the VM controller that synchronizes with ongoing migrations
    // and other attempts to mutate the VM. For the time being, use the instance
    // spec lock to exclude other concurrent attempts to reconfigure this
    // backend.
    let mut spec = vm_controller.instance_spec().await;
```

File: bin/propolis-server/src/lib/server.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/server.rs:1024]: nit: I might push this kind of operational log down into `vcr_replace`, but it's up to you.

File: bin/propolis-server/src/lib/server.rs
Comment (gjcolombo) [inline bin/propolis-server/src/lib/server.rs:1005]: Optional: I've begun evangelizing for putting values into log messages using KV pairs instead of interleaving them into messages:

```suggestion
    slog::info!(
        log,
        "replacing downstairs for Crucible disk";
        "disk_name" => %disk_name,
        "volume_id" => %path_params.id,
        "old_vcr" => %old_vcr_json,
        "new_vcr" => %new_vcr_json,
    );
```

When examining logs you can then filter on the specific fields you're interested in instead of having to search for them in the message body; e.g. you can do `looker -c 'r.volume_id?.contains("<guid>")' and find any record that has a structured volume ID set in it irrespective of how its message was formatted.

Then again, `looker -c 'r.msg.contains("<guid>")` is not *that* much more onerous, so this is your call.

**Review by leftwo — COMMENTED** [review]
I updated the docs too, now that the VCR is expected to be a string.

File: bin/propolis-server/src/lib/server.rs
Comment (gjcolombo) [reply bin/propolis-server/src/lib/server.rs:1022]: Makes sense to me. I suspect we'll want something similar in whatever we ultimately do for #205.

**Review by gjcolombo — APPROVED** [review]
LGTM once we've got a pointer to #205 and a note that `instance_spec()` returns a lock guard in `instance_issue_crucible_vcr_request`. (We touched base on this offline.)

Thanks for driving this one home!

File: bin/propolis-server/src/lib/server.rs
Comment (leftwo) [reply bin/propolis-server/src/lib/server.rs]: I took the log line out, but I'll remember that trick for future logging.

File: bin/propolis-server/src/lib/server.rs
Comment (leftwo) [reply bin/propolis-server/src/lib/server.rs:984]: Got it, I finally figured out what github was showing me here.

**Pull request merged**
